### PR TITLE
[binding] remove binding when corresponding fabric removed

### PR DESCRIPTION
**Pull request by gjc13**
#### Problem

Binding table entries are not correctly removed when fabric is removed.

#### Change overview

* Make `FabricTableDelegate` a linked list to support multiple handlers.
* Remove the bindint table entries when corresponding fabric removed.

#### Testing

```
chip-tool pairing ethernet 2 20202021 3840 172.17.0.3 5540
chip-tool binding bind 1 0 1 6 2 1
chip-tool operationalcredentials remove-fabric 1 2 0
```

Can observe binding removed log in the output.

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

File: src/credentials/FabricTable.h
Comment (bzbarsky-apple) [inline src/credentials/FabricTable.h:323]:
```suggestion
    virtual void OnFabricDeletedFromStorage(CompressedFabricId compressedId, FabricIndex fabricIndex) = 0;
```

**Review by bzbarsky-apple — COMMENTED** [review]
This seems pretty reasonable, but also looks like it will conflict with #14303.  I'd really prefer to review this after that lands and this is merged/rebased...

**Comment by gjc13** [comment]
> This seems pretty reasonable, but also looks like it will conflict with #14303. I'd really prefer to review this after that lands and this is merged/rebased...

Since it's a long holiday in Shanghai and it will be some time before I can address the commentes in #14303, maybe we can merge this first? I'll rebase #14303 after I'm back in office next Monday. @bzbarsky-apple

**Comment by bzbarsky-apple** [comment]
We can do that, but I thought 14303 was pretty close.... and importantly it will need a wholesale re-review on top of these changes.  Which will delay it landing even more....  But whichever you prefer.

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**
